networking/notes/shelley/CoreDIF.md

```diff
@@ -0,0 +1,153 @@
+  block goes out earlier enough that it reaches the next slot leader before
+  it's too late.
+  Always going for worst case actually lowers best case throughput (i.e. tps).
+
+- Minimal spanning trees w.r.t. delta-Q.
+- Fast converging spanning tree algorithms.
+  - We don't have topology change problems?
+  - These cause you to end up with loops which is bad for resource use.
+  - Don't have problems because we have state and so can deal with loops.
+- If we choose a new route, do we create a partition? Can we reason about
+  that so as to prove that we don't create partitions?
+- Neil will think about it.
+  Looking for the right piece of literature about this.
+
+- Make a best guess about a route: you know all of the pairwise delta-Q.
+  You pick (some deterministically, some randomly) who to send it to in the
+  hope that, according to the delta-Q's you know, it will optimally go
+  throughout the network.
+
+- Matrix of delta-Q's (100 x 100). How to fill it up? We know our own, we
+  propagate our own throughout the network (must propagate in order to
+  propagate, hm...)
+  I'm about to publish. I look at the matrix and I can compute paths which
+  with some certainty will get the data to everybody within some time. Of
+  course, it depends upon what the forwarders do.
+  Minimize maximum delay.
+  Also, due to paranoia, send to a few others with some random choice.
+
+- How long before we time-out a peer i.e. set its delta-Q to _|_ ?
+
+- Complication: some of the nodes are relays, some are stakeholders.
+  One network entity per public key????? We need to identify the nodes.
+  Would it not work for a core node to simply claim that its addresses is
+  that of a relay that it controls?
+
+### Unrelated notes
+
+- Benchmarks for deserializing: include in CI?
+- We want numbers to ensure that our logic layer time expenses fit within
+  the time budgets required for diffusion RE slot time.
+- It will be much easier to gather these measurements once we cut out networking
+  from logic.
+
+TIME is such an issue. Everything is limited by time and it causes so much
+trouble.
+Seeing the same issue when thinking about real-time music composition and
+performance. It's hard enough for a single computer to make time guarantees,
+let alone a network of them.
+
+## Questions after meeting with Neil and Duncan November 8
+
+1. Nodes in the core DIF never hold coins? Only stake? Why? Does this buy any
+   security at all? I suppose the machine itself is a target, and you wouldn't
+   want to keep the private key to the _coins_ on that machine. Keeping the
+   key to the delegate is fine, because the holder of the coins can revoke if
+   there's a compromise. Correct?
+
+2. Ability to deny being delegated to? Is this a must have? I would hope so.
+
+3. What does sharing "on the blockchain" really mean? How does it work? I'm
+   very skeptical. Why not flood the (non-core-dif) network with your bearer
+   address encrypted with the target's public key?
```

Comment:
@dcoutts said

> > What does sharing "on the blockchain" really mean? How does it work?
> 
> Its much like sending out an MPC message which gets incorporated into the blockchain. So it's just a special kind of blockchain entry.
> 
> > Why not flood the (non-core-dif) network with your bearer address encrypted with the target's public key?
> 
> It means we have evidence of non-compliance. We could do it by broadcast but then we have no objective evidence later that someone didn't do what they were supposed to. Perhaps for v1 that is ok since we have problems gathering evidence for any of the other bits either.

@georgeee said

> What kind of evidence are you talking about?
> 
> Anyway, also vote for broadcasting.
> When one wanna join Core DIF, he broadcasts message he wants to get addresses of DIF relay nodes and retrieves these encrypted in some way. Then through DIF relay nodes he can authenticate to finally join DIF. See no need, room or rationale to do it on chain.